### ISSUE TITLE
thenables in object literals erroneously dropped with accompanying async/await/Promise

### ISSUE THREAD

**Issue by kzc** [rollup/rollup#4102]
- Rollup Version: rollup v2.50.3
- Operating System (or Browser): n/a
- Node Version (if applicable): v16.1.0

This is an obscure corner of the ECMAScript specification - thenables are similar to Promises:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve#resolving_thenables_and_throwing_errors

### Repro

[2.50.3 REPL](https://rollupjs.org/repl/?version=2.50.3&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwZWZmZWN0cyUyMCUzRCUyMCU1QiU1RCUyQyUyMGVmZmVjdCUyMCUzRCUyMHglMjAlM0QlM0UlMjBlZmZlY3RzLnB1c2goeCklM0IlNUNuUHJvbWlzZS5yZXNvbHZlKCU3QiU1Q24lMjAlMjAlMjAlMjB0aGVuKCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwZWZmZWN0KDEpJTNCJTVDbiUyMCUyMCUyMCUyMCU3RCU1Q24lN0QpJTNCJTVDbihhc3luYyUyMGZ1bmN0aW9uKCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwcmV0dXJuJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHRoZW4oKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBlZmZlY3QoMiklM0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTdEJTVDbiUyMCUyMCUyMCUyMCU3RCUzQiU1Q24lN0QpKCklM0IlNUNuKGFzeW5jJTIwZnVuY3Rpb24oKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBhd2FpdCUyMCglN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwdGhlbiUzQSUyMGZ1bmN0aW9uKCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwZWZmZWN0KDMpJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCU3RCU1Q24lMjAlMjAlMjAlMjAlN0QpJTNCJTVDbiU3RCkoKSUzQiU1Q24oYXN5bmMlMjBmdW5jdGlvbigpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMGF3YWl0JTIwKCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBnZXQlMjB0aGVuKCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwZWZmZWN0KDQpJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHJldHVybiUyMCgpJTIwJTNEJTNFJTIwZWZmZWN0KDUpJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCU3RCU1Q24lMjAlMjAlMjAlMjAlN0QpJTNCJTVDbiU3RCkoKSUzQiU1Q24oYXN5bmMlMjBmdW5jdGlvbigpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMGF3YWl0JTIwKCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjB0aGVuKCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwZWZmZWN0KDYpJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCU3RCU1Q24lMjAlMjAlMjAlMjAlN0QpJTNCJTVDbiU3RCkoKSUzQiU1Q24oYXN5bmMlMjBmdW5jdGlvbigpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMGF3YWl0JTIwYXdhaXQlMjAoJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMHRoZW4ocmVzb2x2ZSklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwcmVzb2x2ZSglN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwdGhlbigpJTIwJTdCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMGVmZmVjdCg3KSUzQiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlN0QlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTdEKSUzQiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlN0QlNUNuJTIwJTIwJTIwJTIwJTdEKSUzQiU1Q24lN0QpKCklM0IlNUNuYXN5bmMlMjBmdW5jdGlvbiUyMGFjdGlvbih4KSUyMCU3QiU1Q24lMjAlMjAlMjAlMjByZXR1cm4lMjB4JTNCJTVDbiU3RCU1Q25hY3Rpb24oJTdCJTdEKSUzQiUyMCUyRiUyRiUyMG5vJTIwc2lkZSUyMGVmZmVjdHMlMjAtJTIwbWF5JTIwYmUlMjBkcm9wcGVkJTVDbnZhciUyMHByb21pc2UlMjAlM0QlMjBhY3Rpb24oOCklM0IlNUNuYWN0aW9uKCU3QiU1Q24lMjAlMjAlMjAlMjB0aGVuKHN1Y2Nlc3MlMkMlMjBmYWlsKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjBzdWNjZXNzKGVmZmVjdCg5KSklM0IlNUNuJTIwJTIwJTIwJTIwJTdEJTVDbiU3RCklM0IlNUNucHJvbWlzZS50aGVuKHglMjAlM0QlM0UlMjBlZmZlY3QoeCkpJTNCJTVDbmFjdGlvbiglN0IlNUNuJTIwJTIwJTIwJTIwdGhlbihyZXNvbHZlKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjByZXNvbHZlKCU3QiU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjB0aGVuKCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwJTIwY29uc29sZS5sb2coZWZmZWN0cy5qb2luKCU1QyUyMiUyMCU1QyUyMikpJTNCJTVDbiUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCUyMCU3RCU1Q24lMjAlMjAlMjAlMjAlMjAlMjAlMjAlMjAlN0QpJTNCJTVDbiUyMCUyMCUyMCUyMCU3RCU1Q24lN0QpJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==)

```
$ cat then.js
const effects = [], effect = x => effects.push(x);
Promise.resolve({
    then() {
        effect(1);
    }
});
(async function() {
    return {
        then() {
            effect(2);
        }
    };
})();
(async function() {
    await ({
        then: function() {
            effect(3);
        }
    });
})();
(async function() {
    await ({
        get then() {
            effect(4);
            return () => effect(5);
        }
    });
})();
(async function() {
    await ({
        then() {
            effect(6);
        }
    });
})();
(async function() {
    await await ({
        then(resolve) {
            resolve({
                then() {
                    effect(7);
                }
            });
        }
    });
})();
async function action(x) {
    return x;
}
action({}); // no side effects - may be dropped
var promise = action(8);
action({
    then(success, fail) {
        success(effect(9));
    }
});
promise.then(x => effect(x));
action({
    then(resolve) {
        resolve({
            then() {
                console.log(effects.join(" "));
            }
        });
    }
});
```

### Expected Behavior

```
$ cat then.js | node
4 1 2 3 5 6 9 8 7
```

### Actual Behavior

```
$ cat then.js | rollup --silent | node
$ 
```
```
$ cat then.js | rollup --silent
const effects = [], effect = x => effects.push(x);
async function action(x) {
    return x;
}
var promise = action(8);
promise.then(x => effect(x));
```

**Edit 2:** test case was further refined.

**Comment by lukastaegert** [comment]
That is unfortunately a known issue, but good to track here nonetheless.

**Comment by kzc** [comment]
Sorry, I had missed #2915.  This issue can be closed as a dup if you like.

I *think* the only scenarios of concern are anything that could create a Promise like `async` `return`, `await` and of course `Promise` itself and its methods. Couldn't the recently introduced object mutation tracker handle something like this? If a `then` property is seen that happens to be of type `function`, always assume it induces a side effect in the aforementioned cases?

**Comment by lukastaegert** [comment]
I actually closed the other one as your tests are much more comprehensive.

**Comment by lukastaegert** [comment]
> If a then property is seen that happens to be of type function, always assume it induces a side effect in the aforementioned cases?

That is actually tricky i.e. we cannot easily "scan" an object for properties from another node. A proper fix would likely mean introducing a new kind of effect (e.g. "hasEffectWhenAwaited"), at which point we might likely want to rework the effect system to work more like deoptimizations with an "event" parameter ("hasEffectWhenIterated" would be something else which would be nice to have so that we can eventually remove for-of loops). Will need to think about this.

**Comment by kzc** [comment]
As for the example in #2915:
```js
var count = 0;
var thenable = Object.defineProperty({}, 'then', { get: function () { count += 1; } });
Promise.resolve(thenable);
assert.equal(count, 1);
```
if `Promise.resolve` has an argument of unknown composition such as the result of `Object.defineProperty()` it ought to assume the worst - that it might contain a `then` property and retain the call.

**Comment by lukastaegert** [comment]
Fix at #4115
